ops/testing.py

```diff
@@ -2586,6 +2606,118 @@ class NonAbsolutePathError(Exception):
+    """
+
+    @abstractmethod
+    def pull(self, path: StrOrPath, *,
+             encoding: Optional[str] = 'utf-8') -> Union[BinaryIO, TextIO]:
+        """Read a file's content from the remote system.
+
+        See :meth:`model.Container.pull` for details.
+        """
+
+    @abstractmethod
+    def push(self,
+             path: StrOrPath,
+             source: Union[bytes, str, BinaryIO, TextIO],
+             *,
+             encoding: str = 'utf-8',
+             make_dirs: bool = False,
+             permissions: Optional[int] = None,
+             user_id: Optional[int] = None,
+             user: Optional[str] = None,
+             group_id: Optional[int] = None,
+             group: Optional[str] = None):
+        """Write content to a given file path on the remote system.
+
+        See :meth:`model.Container.push` for details.
+        """
+
+    @abstractmethod
+    def list_files(self, path: StrOrPath, *, pattern: Optional[str] = None,
+                   itself: bool = False) -> List['FileInfo']:
+        """Return list of directory entries from given path on remote system.
+
+        See :meth:`model.Container.list_files` for details.
+        """
+
+    @abstractmethod
+    def push_path(self,
+                  source_path: Union[StrOrPath, Iterable[StrOrPath]],
+                  dest_dir: StrOrPath):
+        """Recursively push a local path or files to the remote system.
+
+        See :meth:`model.Container.push_path` for details.
+        """
+
+    @abstractmethod
+    def pull_path(self,
+                  source_path: Union[StrOrPath, Iterable[StrOrPath]],
+                  dest_dir: StrOrPath):
+        """Recursively pull a remote path or files to the local system.
+
+        See :meth:`model.Container.pull_path` for details.
+        """
+
+    @abstractmethod
+    def exists(self, path: str) -> bool:
+        """Return true if the path exists on the container filesystem.
+
+        See :meth:`model.Container.exists` for details.
+        """
+
+    @abstractmethod
+    def isdir(self, path: str) -> bool:
+        """Return true if a directory exists at the given path on the container filesystem.
+
+        See :meth:`model.Container.isdir` for details.
+        """
+
+    @abstractmethod
+    def make_dir(
+            self, path: str, *, make_parents: bool = False, permissions: Optional[int] = None,
+            user_id: Optional[int] = None, user: Optional[str] = None,
+            group_id: Optional[int] = None, group: Optional[str] = None):
+        """Create a directory on the remote system with the given attributes.
+
+        See :meth:`model.Container.make_dir` for details.
+        """
+
+    @abstractmethod
+    def remove_path(self, path: str, *, recursive: bool = False):
+        """Remove a file or directory on the remote system.
+
+        See :meth:`model.Container.remove_path` for details.
+        """
+
+
+class _TestingContainerFilesystemProxy:
```

Comment:
Naming nit: Let's drop the `Testing` here, and just make this `_ContainerFilesystemProxy` to more closely relate to `ContainerFilesystem`.

Reply:
Renamed, thanks!